scopesim/source/source_utils.py

```diff
@@ -1,163 +1,219 @@
+    **kwargs : TYPE
+        DESCRIPTION.
 
-    elif lam is not None and len(spectra.shape) == 1 and \
-            isinstance(spectra, np.ndarray) and \
-            isinstance(lam, np.ndarray):
-        spec = SourceSpectrum(Empirical1D, points=lam, lookup_table=spectra)
-        spectra_list += [spec]
+    Raises
+    ------
+    TypeError
+        Raised if an image isn't a FITS HDU or a table isn't an astropy Table.
+    ValueError
+        Raised if a table does not contain the minimum required columns.
 
-    elif ((isinstance(spectra, np.ndarray) and
-           len(spectra.shape) == 2) or
-          (isinstance(spectra, (list, tuple)) and
-           isinstance(spectra[0], np.ndarray))) and \
-            isinstance(lam, np.ndarray):
+    Returns
+    -------
+    None
 
-        for sp in spectra:
-            spec = SourceSpectrum(Empirical1D, points=lam, lookup_table=sp)
-            spectra_list += [spec]
+    """
+    if (filename := kwargs.get("filename")) is not None:
+        if find_file(filename) is None:
+            logger.warning("filename was not found: %s", filename)
 
-    return spectra_list
+    if (image_hdu := kwargs.get("image")) is not None:
+        if not isinstance(image_hdu, (fits.PrimaryHDU, fits.ImageHDU)):
+            raise TypeError(
+                f"Image must be fits.HDU object: {type(image_hdu) = }")
 
+        if not wcs.find_all_wcs(image_hdu.header):
+            logger.warning(
+                "Image does not contain valid WCS. %s", wcs.WCS(image_hdu))
 
-def photons_in_range(spectra, wave_min, wave_max, area=None, bandpass=None):
+    if (tbl := kwargs.get("table")) is not None:
+        if not isinstance(tbl, Table):
+            raise TypeError(
+                f"Table must be astropy.Table object: {type(tbl) = }")
+
+        if not {"x", "y", "ref"}.issubset(tbl.colnames):
+            raise ValueError(
+                "Table must contain at least the following column names: 'x', "
+                f"""'y', 'ref'; found only: '{"', '".join(tbl.colnames)}'""")
+            # TODO py312: The triple quotes will become redundant in 3.12 !
+
+
+def convert_to_list_of_spectra(spectra, lam) -> list[SourceSpectrum]:
+    """Produce SourceSpectrum instances or pass them through."""
+    def _synphotify(spec):
+        if not isinstance(lam, np.ndarray):
+            raise TypeError("If spectra is/are given as array(s), lam must be "
+                            "an array as well.")
+        return SourceSpectrum(Empirical1D, points=lam, lookup_table=spec)
+
+    def _from_arrays(specarrays):
+        for spec in specarrays:
+            yield _synphotify(spec)
+
+    def _get_list():
+        if isinstance(spectra, SourceSpectrum):
+            yield spectra
+            return
+
+        if (isinstance(spectra, Iterable) and
+                not isinstance(spectra, np.ndarray)):
+            if all(isinstance(spec, SourceSpectrum) for spec in spectra):
+                yield from spectra
```

Comment:
I don't really like this refactoring, because the code is more complex and the yield statements indicate that it works with generators but it does not, because the `all()` would have used up everything. E.g.

```python
from itertools import count, takewhile

numbers = takewhile(lambda x: x< 20, count(10))

def convert(spectra):
    def get_list():
        if all(isinstance(spec, int) for spec in spectra):
            yield from spectra

    return list(get_list())

list_of_numbers = convert(numbers)

print(f"{list_of_numbers=}")
```
will print
```
list_of_numbers=[]
```
I think it is better to simplify this by converting the input to a list and remove the yields.

Reply:
Added list conversion in the iterable case, so this should now be fine. Perhaps find a better implementation in the future.